### ISSUE TITLE
V8: Fix A & AA accessibility issues throughout the back-office

### ISSUE THREAD

**Issue by MMasey** [umbraco/Umbraco-CMS#5277]
# IMPORTANT UPDATE
We have an updated trello board with all the tickets from the latest Accessibility Audit.  
Trello Board for Umbraco Accessibility Issues (8.4>) - https://trello.com/b/7M8RosrE/umbraco-accessiblity-issues-84

Note - I've left the link for the old trello board for reference. We may close this ticket soon and open a new one to tidy things up a little bit.

---

There are currently a large number of accessibility and in tern usability issues throughout the back-office. One of the wonderful members of the PR team Danny Lancaster has created an accessibility audit, detailing each of the bugs found. 

There are currently **151** known issues that need to be fixed. Rather than create an issue for each one, the idea is to do the following.

1. Have a look through the table below and pick an issue that you'd like to fix
2. Fix the issue 😄
3. Submit a PR
4. Leave a comment with a link to your PR here and we will update the trello board (see comment below).

I've added some links to various articles and tools to help you on your way below. https://github.com/umbraco/Umbraco-CMS/issues/5277#issuecomment-483609542

If you have any questions you can reach out either here or on the Umbraco Community #accessibility Slack channel. You can join the community slack channel by going to [umbracians.chat](http://umbracians.chat/).

**We have a Trello board with a list of resources and goals available at [trello.com/b/MwD8xuz3/umbraco-accessibility](https://trello.com/b/MwD8xuz3/umbraco-accessibility)**

**Comment by MMasey** [comment]
To make things easier I've moved the spreadsheet over to here 😄

![image](https://user-images.githubusercontent.com/6573613/56197140-19db2600-6030-11e9-9f91-e2e7f770b1f3.png) |  | Umbraco V8 Accessibility Audit  |   |  
-- | -- | -- | -- | --
Issue | Level | Description | Area | PR Link
1 | Level AA : 2.4.7 - Unclear focus area | Tab states (focus) is insufficient or missing | General |  
2 | Level A : 2.4.1 - Skip links | No "skip navigation" links present | General |  
3 | Level AA : 1.4.4 - Resizing text | Fonts do not resize using the browser font size settings | General |  
4 | Level AA : 2.4.7 - Unclear focus area | Bottom publishing options ('Preview', 'Save', 'Save and publish') have insufficient focus (tab) state | General |  
5 | Level A : 2.1.1 - Keyboard and screen reader | Bottom publishing option "Save and publish" has a drop down that can't be accessed via keyboard | General |  
6 | Level A : 2.1.1 - Keyboard and screen reader | Unable to add child items to content tree (left hand) via keyboard | General |  
7 | Level A : 2.1.1 - Keyboard and screen reader | Unable to open 'links' options (create etc.) using keyboard | General |  
8 | Level A : 2.4.4 - Unclear link/button text | "Actions" is identified as a button, but behaves like a drop down - screen readers | General |  
9 | Level A : 2.4.4 - Unclear link/button text | "Actions" drop down options do not give context to the user. E.g. "Delete" - screen readers | General |  
 
10 | Level A : 2.4.3 - Illogical focus order | Should receive focus as first item on page (when open) | Welcome to Umbraco (modal) |  
11 | Level A : 2.1.1 - Keyboard and screen reader | Unable to navigate to close 'X' icon with keyboard | Welcome to Umbraco (modal) |  
12 | Level A : 2.1.2 - Keyboard trap | Modal is missing keyboard trap | Welcome to Umbraco (modal) |  
 
13 | Level AA : 2.4.7 - Unclear focus area | Menu items have insufficient focus (tab) states | Top menu (Umbraco sections) |  
14 | Level A : 2.1.1 - Keyboard and screen reader | Unable to close search once open | Top menu (Umbraco sections) |  
15 | Level AA : 2.4.7 - Unclear focus area | No focus state present for the user icon (top right hand side) | Top menu (Umbraco sections) | https://github.com/umbraco/Umbraco-CMS/pull/5249
16 | Level A : 2.4.4 - Unclear link/button text | Search icon and help (?) icon both identify as "link" - no context for screen readers | Top menu (Umbraco sections) |  
17 | Level A : 2.4.4 - Unclear link/button text | User account icon requires ARIA-hidden to give context to users - screen readers | Top menu (Umbraco sections) |  
18 | Level A : 2.4.4 - Unclear link/button text | Does not identify if child items are present for drop down - screen readers | Top menu (Umbraco sections) |  
 
19 | Level AA : 2.4.7 - Unclear focus area | No focus (tab) state present for the content tree items | Left hand menu (content tree) |  
20 | Level A : 2.1.1 - Keyboard and screen reader | User can interact with content tree item, but is unable to open child items when applicable | Left hand menu (content tree) |  
21 | Level A : 2.4.4 - Unclear link/button text | Content link identified as "list with 1 item" but does not identify list for rest of content tree - screen readers | Left hand menu (content tree) |  
22 | Level A : 2.4.4 - Unclear link/button text | When 'DO SOMETHING ELSE' is interacted with (requires mouse to operate) the confirmation buttons need to give context - screen readers | Left hand menu (content tree) |  
 
23 | Level AA : 1.4.3 - Insufficient colour contrast | Green buttons present (e.g. "Free 14 days trial of Umbraco 8 >>") has insufficient colour contrast (Fails AA + AAA) | Information (Content) |  
24 | Level AA : 1.4.3 - Insufficient colour contrast | Footer links (e.g. "Documentation", "Community" etc.) has insufficient colour contrast (Fails AA + AAA) | Information (Content) |  
25 | Level A : 2.4.4 - Unclear link/button text | Buttons with icon '>>' needs ARIA role as currently identifies as "Greater greater link" - screen readers | Information (Content) |  
 
26 | Level AA : 2.4.7 - Unclear focus area | No focus (tab) states present for form fields | Login |  
27 | Level A : 2.4.4 - Unclear link/button text | Errors are not identified for screen reader users | Login |  
28 | Level A : 2.4.3 - Illogical focus order | Show/hide password should focus user back to password field on tab (need to reverse tab) | Login |  
 
29 | Level A : 2.1.1 - Keyboard and screen reader | Unable to navigate to 'Content' or 'Info' or drop down using keyboard (only 'Actions') | Home (node) |  
30 | Level AA : 2.4.7 - Unclear focus area | No focus (tab) states are present for any of the form fields | Home (node) |  
31 | Level A : 2.4.3 - Illogical focus order | When '+' interacted with to add more sections, focus should go to the top of that section | Home (node) |  
32 | Level A : 2.4.4 - Unclear link/button text | "+" icon is identified as "link" - screen readers | Home (node) |  
33 | Level A : 2.1.1 - Keyboard and screen reader | When '+' has been interacted with, the item section controls (pencil, move, bin) can not be navigated to via keyboard | Home (node) |  
34 | Level A : 2.1.1 - Keyboard and screen reader | Links components (e.g. 'Footer links') can not be tabbed to ('Edit', 'Remove') | Home (node) |  
35 | Level A : 2.4.3 - Illogical focus order | Links components (e.g. 'Footer links'), when 'Add' interacted with, focus should goes to the top of that overlay | Home (node) |  
36 | Level A : 2.1.2 - Keyboard trap | Links components (e.g. 'Footer links'), when 'Add' interacted with, overlay section should have keyboard trap | Home (node) |  
37 | Level A : 2.4.4 - Unclear link/button text | Links components (e.g. 'Footer links'), 'Add' is identfied as "Add link" but doesn't identify the label - screen readers | Home (node) |  
38 | Level A : 2.1.1 - Keyboard and screen reader | Meta keywords, more can be added (by typing and hitting enter), but unable to remove/interact with existing labels | Home (node) |  
39 | Level A : 1.3.1 - Fieldset missing legend | Only identifies the placeholder text, not the form label - screen readers | Home (node) |  
40 | Level AA : 2.4.7 - Unclear focus area | Toggle options have insufficient focus (tab) state | Home (node) |  
41 | Level A : 2.4.4 - Unclear link/button text | Toggle options are identified as "button" but behaves like a toggle option | Home (node) |  
42 | Level A : 1.3.1 - Fieldset missing legend | Toggle options are identified as "button" but does not identifiy label - screen readers | Home (node) |  
43 | Level A : 1.3.1 - Fieldset missing legend | "Enter a name" for page title, should be "Enter a page name" - screen readers | Home (node) |  
44 | Level A : 2.4.4 - Unclear link/button text | "Actions" is identified as a button, but behaves like a drop down - screen readers | Home (node) |  
45 | Level A : 2.4.4 - Unclear link/button text | "Actions" drop down options do not give context to the user. E.g. "Delete" - screen readers | Home (node) |  
46 | Level A : 2.1.1 - Keyboard and screen reader | Once new row has been added to page grid, unable to navigate to using keyboard | Home (node) |  
47 | Level A : 2.1.2 - Keyboard trap | Infinite editing - requires keyboard trap | Home (node) |  
48 | Level A : 2.4.3 - Illogical focus order | Infinite editing - Requires focus when open | Home (node) |  
49 | Level A : 2.4.4 - Unclear link/button text | Infinite editing - Add image - '+' in breadcrumb is identified as "link" - screen readers | Home (node) |  
50 | Level A : 2.4.4 - Unclear link/button text | Infinite editing - Add image - Search does not give context to the user - screen readers | Home (node) |  
51 | Level A : 2.1.1 - Keyboard and screen reader | Infinite editing - Add image - unable to navigate to already uploaded images using keyboard | Home (node) |  
52 | Level AA : 2.4.7 - Unclear focus area | Infinite editing - Add image - Upload doesn't have a very clear focus (tab) state | Home (node) |  
 
53 | Level A : 2.1.1 - Keyboard and screen reader | Unable to navigate to 'Main Image' using a keyboard | About (content tree) |  
54 | Level A : 2.1.1 - Keyboard and screen reader | Able to interact with '+' to bring up new grid line, but unable to interact with using keyboard | About (content tree) |  
55 | Level A : 2.1.1 - Keyboard and screen reader | Once grid line option has been selected (needs to be selected using a mouse), unable to interact with the grid options uisng keyboard | About (content tree) |  
56 | Level A : 2.4.3 - Illogical focus order | Once grid options has been selected (needs to be selected using a mouse), focus should go the grid options modal | About (content tree) |  
57 | Level A : 2.1.2 - Keyboard trap | Once grid options has been selected (needs to be selected using a mouse), grid options modal requires keyboard trap | About (content tree) |  
58 | Level A : 2.4.4 - Unclear link/button text | Reorder' button does not give context to the user (what are they reordering) - screen readers | About (content tree) |  
59 | Level A : 2.4.4 - Unclear link/button text | Reorder' button requires an ARIA-ALERT to identify when the options change to "I am done reordering" - screen readers | About (content tree) |  
60 | Level A : 2.4.4 - Unclear link/button text | "+" icon is identified as "link" - screen readers | About (content tree) |  
 
61 | Level A : 2.1.1 - Keyboard and screen reader | Create article' button can't be accessed via keyboard | Blog (content tree) |  
62 | Level A : 2.1.1 - Keyboard and screen reader | Search grid option drop down can't be accessed via keyboard | Blog (content tree) |  
63 | Level A : 2.1.1 - Keyboard and screen reader | Existing blogs can not be accessed via keyboard | Blog (content tree) |  
 
64 | Level A : 2.1.1 - Keyboard and screen reader | Meta keywords, more can be added (by typing and hitting enter), but unable to remove/interact with existing labels | Contact (content tree) |  
65 | Level A : 1.3.1 - Fieldset missing legend | Only identifies the placeholder text, not the form label - screen readers | Contact (content tree) |  
66 | Level AA : 2.4.7 - Unclear focus area | Toggle options have insufficient focus (tab) state | Contact (content tree) |  
67 | Level A : 2.4.4 - Unclear link/button text | Toggle options are identified as "button" but behaves like a toggle option | Contact (content tree) |  
68 | Level A : 1.3.1 - Fieldset missing legend | Toggle options are identified as "button" but does not identifiy label - screen readers | Contact (content tree) |  
69 | Level A : 1.3.1 - Fieldset missing legend | "Enter a name" for page title, should be "Enter a page name" - screen readers | Contact (content tree) |  
 
70 | Level A : 2.1.1 - Keyboard and screen reader | Meta keywords, more can be added (by typing and hitting enter), but unable to remove/interact with existing labels | Search (content tree) |  
71 | Level A : 1.3.1 - Fieldset missing legend | Only identifies the placeholder text, not the form label - screen readers | Search (content tree) |  
72 | Level AA : 2.4.7 - Unclear focus area | Toggle options have insufficient focus (tab) state | Search (content tree) |  
73 | Level A : 2.4.4 - Unclear link/button text | Toggle options are identified as "button" but behaves like a toggle option | Search (content tree) |  
74 | Level A : 1.3.1 - Fieldset missing legend | Toggle options are identified as "button" but does not identifiy label - screen readers | Search (content tree) |  
75 | Level A : 1.3.1 - Fieldset missing legend | "Enter a name" for page title, should be "Enter a page name" - screen readers | Search (content tree) |  
 
76 | Level A : 2.1.1 - Keyboard and screen reader | Meta keywords, more can be added (by typing and hitting enter), but unable to remove/interact with existing labels | Links (content tree) |  
77 | Level A : 1.3.1 - Fieldset missing legend | Only identifies the placeholder text, not the form label - screen readers | Links (content tree) |  
78 | Level AA : 2.4.7 - Unclear focus area | Toggle options have insufficient focus (tab) state | Links (content tree) |  
79 | Level A : 2.4.4 - Unclear link/button text | Toggle options are identified as "button" but behaves like a toggle option | Links (content tree) |  
80 | Level A : 1.3.1 - Fieldset missing legend | Toggle options are identified as "button" but does not identifiy label - screen readers | Links (content tree) |  
81 | Level A : 1.3.1 - Fieldset missing legend | "Enter a name" for page title, should be "Enter a page name" - screen readers | Links (content tree) |  
 
82 | Level A : 2.1.1 - Keyboard and screen reader | Grid filter icon drop down can't be accessed via keyboard | Recycle bin (content tree) |  
83 | Level A : 2.1.1 - Keyboard and screen reader | Unable to navigate to previously deleted items | Recycle bin (content tree) |  
84 | Level A : 2.4.4 - Unclear link/button text | Does not advise which/how many of previously uploaded media is selected (needs to be selected using a mouse) | Recycle bin (content tree) |  
85 | Level A : 2.4.4 - Unclear link/button text | Once selected, user can then tab to additional options but does not give context to the user - screen readers | Recycle bin (content tree) |  
 
86 | Level A : 2.1.1 - Keyboard and screen reader | Create' drop down can't be accessed via keyboard | Media (top menu) |  
87 | Level A : 2.1.1 - Keyboard and screen reader | Grid filter icon drop down can't be accessed via keyboard | Media (top menu) |  
88 | Level A : 2.1.1 - Keyboard and screen reader | Media uploader widget can't be accessed via keyboard | Media (top menu) |  
89 | Level A : 2.1.1 - Keyboard and screen reader | Previously upload media can't be accessed via keyboard (only on left hand menu) | Media (top menu) |  
90 | Level A : 2.4.4 - Unclear link/button text | Does not advise which/how many of previously uploaded media is selected (needs to be selected using a mouse) | Media (top menu) |  
91 | Level A : 2.4.4 - Unclear link/button text | Once selected, user can then tab to additional options but does not give context to the user - screen readers | Media (top menu) |  
92 | Level A : 2.4.3 - Illogical focus order | When left hand menu interacted with (by hitting enter), focus should go to open media settings | Media (top menu) |  
93 | Level A : 1.1.1 - Alternative text | Open media - no option to add alternative text here? | Media (top menu) |  
94 | Level A : 2.4.4 - Unclear link/button text | Open media - "Remove files" can be tabbed to, but does not give context to the user - screen readers | Media (top menu) |  
 
95 | Level A : 2.1.1 - Keyboard and screen reader | Can't access any of the secondary menu items (Welcome, Examine Management etc.) via keyboard | Settings (top menu) |  
96 | Level A : 2.4.4 - Unclear link/button text | Examine Management - needs to give context - screen readers | Settings (top menu) |  
97 | Level A : 2.4.4 - Unclear link/button text | Published Status - needs to give context - screen readers | Settings (top menu) |  
98 | Level A : 2.4.4 - Unclear link/button text | Models Builder - Reload button needs to identify help tip to give context - screen readers | Settings (top menu) |  
99 | Level A : 2.1.1 - Keyboard and screen reader | Health Check - unable to navigate to additional options (Configuration, live environment etc.) via keyboard | Settings (top menu) |  
 
100 | Level A : 2.1.1 - Keyboard and screen reader | Secondary menu items (Packages, installed, install local etc.) can't be accessed via keyboard | Packages (top menu) |  
101 | Level A : 2.4.3 - Illogical focus order | When filter option (Collaborations, backoffice extensions etc.) has been interacted with, focus should go to options | Packages (top menu) |  
102 | Level A : 2.1.1 - Keyboard and screen reader | Screen reader only identifies the description text present on screen. Requires ARIA role to read full description and not amount of 'likes' | Packages (top menu) |  
103 | Level A : 2.1.1 - Keyboard and screen reader | Should identify section labels/headers when tabbing into new section - "New releases" - Screen readers | Packages (top menu) |  
104 | Level A : 2.1.1 - Keyboard and screen reader | Open Package - Identifies back button as "Left hand arrow back button" - screen readers | Packages (top menu) |  
105 | Level A : 2.1.1 - Keyboard and screen reader | Open Package - Can only tab to 'Back' and 'Install package' button - missing important content in package description (with interactive elements) | Packages (top menu) |  
 
106 | Level AA : 2.4.7 - Unclear focus area | "Create user" doesn't have a very clear focus (tab) state | Users (top menu) |  
107 | Level A : 2.1.1 - Keyboard and screen reader | Unable to navigate to search filter icon drop down using keyboard | Users (top menu) |  
108 | Level A : 2.1.1 - Keyboard and screen reader | Unable to navigate to 'Groups' icon filter | Users (top menu) |  
109 | Level A : 2.1.1 - Keyboard and screen reader | Filter users drop downs do not identify as drop downs (open/close) - screen readers | Users (top menu) |  
110 | Level A : 2.1.1 - Keyboard and screen reader | Unable to navigate to user cards via keyboard | Users (top menu) |  
111 | Level A : 2.1.1 - Keyboard and screen reader | Open User Details - Back icon arrow is identified as "Umbraco link" - screen readers | Users (top menu) |  
112 | Level A : 1.3.1 - Fieldset missing legend | Open User Details - Language drop down does not identify form label - screen readers | Users (top menu) |  
113 | Level AA : 2.4.7 - Unclear focus area | Open User Details - 'Change password' button does not have sufficient tab state | Users (top menu) |  
114 | Level A : 1.3.1 - Fieldset missing legend | Open User Details - 'Change password' fields are all missing form labels - screen readers | Users (top menu) |  
115 | Level AA : 2.4.7 - Unclear focus area | Open User Details - 'Cancel' button has no focus (tab) state present | Users (top menu) |  
116 | Level AA : 2.4.7 - Unclear focus area | Open User Details - 'Save' button has no focus (tab) state present | Users (top menu) |  
117 | Level A : 2.1.1 - Keyboard and screen reader | Create user - Identifies back button as "Left hand arrow back button" - screen readers | Users (top menu) |  
118 | Level A : 1.3.1 - Fieldset missing legend | Create user - It is identifying the placeholder text, but not the actual form label - screen readers | Users (top menu) |  
119 | Level A : 2.4.4 - Unclear link/button text | Create user - User group 'Add' button doesn't give context to the user - screen readers | Users (top menu) |  
120 | Level A : 3.3.1 - Unclear error messages | Create user - Errors are not identified - screen readers | Users (top menu) |  
121 | Level A : 2.4.3 - Illogical focus order | Create user - Errors should retain focus when submit occurs | Users (top menu) |  
122 | Level A : 1.1.1 - Alternative text | No alt text on any icons | Users (top menu) |  
 
123 | Level AA : 2.4.7 - Unclear focus area | Create members - form fields missing focus (tab) states | Members (top menu) |  
124 | Level A : 2.4.4 - Unclear link/button text | Create members - "Enter name" but should context for username - screen readers | Members (top menu) |  
125 | Level A : 2.4.4 - Unclear link/button text | Create members - 'Is approved' toggle is identified as "button" - screen readers | Members (top menu) |  
126 | Level A : 2.4.4 - Unclear link/button text | Create members - Password fields do not identify with any form label so no context available - screen readers | Members (top menu) |  
127 | Level A : 2.4.4 - Unclear link/button text | Create members - Member group add button doesn't give any context to the user - screen readers | Members (top menu) |  
128 | Level A : 3.3.1 - Unclear error messages | Create members - Errors are not identified - screen readers | Members (top menu) |  
129 | Level A : 2.4.3 - Illogical focus order | Create members - Errors should retain focus when submit occurs | Members (top menu) |  
130 | Level A : 2.4.4 - Unclear link/button text | Create members - 'Password Answer' + 'Password Question' don't appear to do anything? | Members (top menu) |  
131 | Level A : 2.1.1 - Keyboard and screen reader | Create members - Requires ARIA-ALERT when user saved to advise information has updated/user been created - screen readers | Members (top menu) |  
132 | Level AA : 2.4.7 - Unclear focus area | Members list - No focus (tab) states present for any header filters | Members (top menu) |  
133 | Level A : 2.4.4 - Unclear link/button text | Members list - Checkbox has no form label so identifies as "check box" which gives no context - screen readers | Members (top menu) |  
134 | Level A : 2.1.1 - Keyboard and screen reader | Members list - Does not advise if filter is already applied (e.g. sorted by email acsending) - screen readers | Members (top menu) |  
135 | Level AA : 2.4.7 - Unclear focus area | Members list - Existing users have no focus (tab) states present | Members (top menu) |  
 
136 | Level A : 2.1.1 - Keyboard and screen reader | Top icons (Design + Settings) can not be accessed via keyboard | Forms (top menu) |  
137 | Level A : 2.4.4 - Unclear link/button text | Icons in question type (Cog + Bin + lock) are identified as "link" - screen reader | Forms (top menu) |  
138 | Level A : 2.4.4 - Unclear link/button text | ARIA-Hidden required for flow information to give better context to the user - screen reader | Forms (top menu) |  
139 | Level A : 2.4.3 - Illogical focus order | Adding question - Infinite editing - Focus does not go to top layer | Forms (top menu) |  
140 | Level A : 2.1.2 - Keyboard trap | Adding question - Infinite editing - No keyboard trap present | Forms (top menu) |  
141 | Level A : 2.4.3 - Illogical focus order | Adding question - when selection made, focus does not return to flow (goes to top of page) | Forms (top menu) |  
142 | Level A : 1.3.1 - Fieldset missing legend | Add question - 'Default value' + 'placeholder' - missing form label - identified as "Edit blank" - screen reader | Forms (top menu) |  
143 | Level A : 3.3.2 - Form fields missing associated form labels | Add question - 'Validation' drop down is missing form label - "Combo box collapsed" | Forms (top menu) |  
 
144 | Level A : 3.3.2 - Form fields missing associated form labels | When activate - Left hand menu - Top item changes to "English (United States) drop down - identified as "List with 1 item heading level 5" - screen reader | Duel language |  
145 | Level A : 2.4.4 - Unclear link/button text | Language drop downs do not advise as drop downs - screen readers | Duel language |  
146 | Level A : 2.4.4 - Unclear link/button text | Close 'X' icon next to language drop down is identified as "link" - screen reader | Duel language |  
147 | Level A : 2.1.1 - Keyboard and screen reader | Although one side is dimmed out/unable to click in with mouse, the RTE area and the '+' icon can be tabbed to and interacted with | Duel language |  
148 | Level A : 2.1.1 - Keyboard and screen reader | Dimmed content may benefit from ARIA-Hidden role to advise user to use other column to edit both - screen reader | Duel language |  
149 | Level A : 2.1.1 - Keyboard and screen reader | Reorder' button is displaying about title and appearing twice - seems inconsistent | Duel language |  
150 | Level A : 2.1.1 - Keyboard and screen reader | Reorder button is repeated | Duel language |  
151 | Level A : 3.3.2 - Form fields missing associated form labels | Missing many form labels | Duel language |

**Comment by MMasey** [comment]
Here are some links to various tools and reference materiel should you require any help 🙂

**Websites & Articles**

- [Web Content Accessibility Guidelines (WCAG) Overview](https://www.w3.org/WAI/standards-guidelines/wcag/)
- [Getting Comfortable with WCAG](https://seesparkbox.com/foundry/getting_comfortable_with_wcag)
- [Accessibility according to actual people with disabilities](https://axesslab.com/accessibility-according-to-pwd/)
- [Colorblind Accessibility on the Web – Fail and Success Cases](https://axesslab.com/colorblind-accessibility-web-fail-success-cases/)
- [AngularJS guide to accessibility](https://docs.angularjs.org/guide/accessibility)

**Tools**
- [Accessibility Insights - Chrome Extension](https://chrome.google.com/webstore/detail/accessibility-insights-fo/pbjjkligggfmakdaogkfomddhfmpjeni)
- [Axe: Accessibility for web development](https://www.deque.com/axe/)
- [Color Contrast Checker](https://webaim.org/resources/contrastchecker)


**Activate built in screen reader**
- Mac - toggle Voice Over by pressing Command-F5
- PC- toggle Narrator by pressing Winkey +Enter

**Comment by MMasey** [comment]
PR for addressing 26 & 28 - #5304

**Comment by shane-sigma** [comment]
PR for 4 - #5320

**Comment by shane-sigma** [comment]
PR for 11 - #5333

**Comment by shane-sigma** [comment]
PR for 19 - #5334

It's worth noting that many of these focus state bugs are related to there being specific `outline:0` rules in the CSS. 
From an accessibility POV, these should not be present without an alternate focus state/style - which isn't generally present from what I have seen - it would be good to completely remove all of these rules but I am unsure of why they were added in the first place, it's also worth considering any knock-on effect that the removal of these styles might cause.

**Comment by MMasey** [comment]
> PR for 19 - #5334
> 
> It's worth noting that many of these focus state bugs are related to there being specific `outline:0` rules in the CSS.
> From an accessibility POV, these should not be present without an alternate focus state/style - which isn't generally present from what I have seen - it would be good to completely remove all of these rules but I am unsure of why they were added in the first place, it's also worth considering any knock-on effect that the removal of these styles might cause.

Thanks for the great work Shane, as per the comments in your PR, I've updated the table with the PR submitted by @nielslyngsoe for 19.

**Comment by Matthew-Wise** [comment]
PR for 27 #5450

**Comment by MMasey** [comment]
PR for 15, 16, 17 #5544

**Comment by RachBreeze** [comment]
PR for 43, 69, 75, 81 #5585

**Comment by zersiax** [comment]
Hi there :-)
I am a blind developer and recently became interested in Umbraco. I tweeted about some accessibility-related issues I ran into, after which @MMasey  , @ArnoldV  as well as others pointed me to the work being done here to further Umbraco's back office accessibility.
I was asked to share my insights, both as a blind developer as well as an accessibility expert, to indicate what issues were serious showstoppers. Below, I will share my notes on the list of issues listed above.
TL,DR: Most notably, three main areas currently really need improvement before Umbraco is a viable option for blind admins, editors and developers. These areas are textual alternatives for various icons in the interface, semantic information for the various user interface components (is this a button or a checkbox? What does this edit button actually edit?) and inaccessibility of various parts of the interface with the keyboard focus (adding child nodes to the content tree, for example). 
Below, I share my thoughts on the issues that caught my eye. If anyone has questions about these notes, feel free to reach out on Slack or Twitter. I will make sure to keep an eye on both.


Issue 5:
Can you elaborate a bit on what that means? What options are not reachable? Can they be reached or accessed through other means when only using the keyboard? If these are at all important options, not having access to them would be a significant access barrier for screen reader users like myself as there almost literally wouldn't be a way to access these options at all, which pretty much stops admins, editors and developers that rely on these technologies in their tracks.

issue 6:
Is there an alternative to still be able to add content nodes through the keyboard, perhaps by setting a field on the newly created node? If not, this flat out makes Umbraco unusable for it's intended purpose for people relying on screen readers as well as people who can't use the mouse for some reason (RSI, physical disability etc)

Issue 7: That is the hover menu, right? That is a big blocker for keyboard users as well.

issue 8: That isn't so much a case of unclear text as it is a case of missing aria markup, I would say. Adding aria markup to mark this button as expandable/collapsable pretty much fixes this. That is, if I understand the issue correctly; it is an actual button that behaves like a dropdown when clicked? 

issue 14: If escape doesn't close search, then yes that seems accurate. Then again, search appears at the bottom of the DOM, so unless it overlaps another element on the page directly often screen reader users won't even be aware search is open in the first place until they navigate all the way to the bottom of the page (screen readers follow the DOM for the most part when rendering HTML content).


Issue 16: This is part of a larger issue where iconography is the only way to convey important information without textual alternatives. I see this all throughout the Umbraco UI, i.e. some menu options, buttons that change state (checkboxes) etc. and is actually one of the main entry barriers for screen reader users right now. Even a simple thing like following the basic website tutorial makes me have to guess at some things, as I am to " check the allow root" in the permissions tab of a new document type, which to me reads as a button with no text on it.

issue 17: Can you elaborate on this? Aria-hidden is used to hide an item from the screen reader rendering, do you mean the icon should be aria-hidden and a textual sr-only span should be used to convey that this is the user icon?

Issue 20: I am really sorry for the duplicate questions here but I am still very new to Umbraco. Is there another way to get at these children? If not, that is a serious blocker.

Issue 22: sounds like a serious one, but I am not quite sure what it refers to exactly

issue 25: I am unclear how an aria-role would fix this? You want the screen reader to say "Next"  or " Continue"  instead? That would require overriding the value of the control, not it's role.

issue 29: Ditto to other similar issues. If there is no alternative way to access these options, that means it stops keyboard, screen reader and other assistive tech users right there and requires them to either ask someone for help or stop what they were doing. This is not acceptable in a lot of cases, especially in an employment situation.

issue 32: This is the same problem as issue 16. The screen reader says "link" because that is the item's role, but the icon has no textual alternative, and is therefore not spoken.

issue 33: That is another serious one, although in a pinch my suspicion is that a screen reader's virtual cursor commands might be of some use here. Need to test.

issue 41: This is slightly related to issue 8. Controls don't have the right role; in this case that isn't supposed to be a button but a checkbox. You can either use a native HTLM element checkbox and get most of what you need for free, or add role=checkbox, aria-state toggled when the button is interacted with and keyboard handling code using JS. From a coding standpoint, just using a checkbox element would be preferable. However, this might introduce some browser-default styling that needs to be gotten rid of.

issue 42: I think this mainly happens for checkboxes that only have an icon to denote their function. Also this has very little to do with form fieldsets, this is a name, role , value problem.

issue 44: duplicate of issue 8?

issue 46: That is problematic. Would refreshing the page help here? Asking to judge priority of this issue, not stating that that would be an acceptable workaround :-)

issue 49: Same problem as 32 and therefore same as 16.

Issue 51: That mainly breaks the non-happy path, e.g. user uploaded the wrong image and now can't get rid of it/change it.

Issue 54: Is that the same issue as number 46? Same question about issue 55.

issue 58: This is another pattern I see a lot as it were. It is not always clear to me, for example in the document type dialog, if I am editing a group, a property, what property I am working with etc. which makes that part of the interface very finicky to work with. E.g. I start changing editors for the wrong fields etc. This isn't as debilitating as some of the other issues but it is rather frustrating.

issue 60: same as 16, 32 and 49

Issue 61: Is there a keyboard shortcut to activate this button instead? If not, that essentially means keyboard users can't write articles on the blog part. Serious blocker.

	issue 67: same problem as issue 41, ditto for 73, 79

Sounds to me like the recycle bin pretty much just isn't usable at all using a keyboard / screen reader right now.

Issue 88: That basically means creators can't use media in their articles when they don't use a mouse. Serious showstopper.

Issue 93: That is an interesting one. This is a big problem for people who want to create accessible content using Umbraco, which pretty much means it can't easily be used for public sector-based development, as not being accessible can have serious consequences in that space especially in the United States.

issue 99: These options aren't reachable at all? Same question for 100.

issue 102: This is not something an aria role can fix. An aria-labeledby perhaps.

issue 105: That is why screen readers don't just tab through everything :-) IF there is textual content around interactive elements, it is up to the screen reader user to use alternative methods to read that information. Clear link text still applies, but static text does not need to be made tab-reachable here

issue 111: same as 60, 49, 32 and 16. That is happening because there is just no meaningful content to report.

issue 117: Seen this one pop up a few times in the list and figured I'd comment, that has to do with the fact that button has a unicode character on it that is called Left hand arrow". Probably this one, or a variation on it: ←   

issue 119: Haven't pointed them all out to keep this text somewhat within readable lengths, but this is a similar problem to 58 and there's a few more on this list like it.

issue 122: I like this, this is essentially what I've been saying in regards to 16, 32 and friends. However, I think the icons here are not actual img tags, so an alt attribute wouldn't work here. You will probably want to aria-hide the icon and sr-only a label for them, or figure out some other strategy that makes the textual label accessible. Keep in mind that not just screen readers might want some kind of textual explanation; not everyone knows what a gear  icon is supposed to mean especially if they aren't very computer-literate.

issue 125: we've seen this one before as well. Wrong semantic role. See 41, for example. Again, haven't pointed all of these occurrences out but the pattern should be clear :-)

issue 126: that really is a low-hanging fruit, as well as 133, 142 and 143 :-)

issue 136: You guys are smart. You know what I am going to ask ;)

issue 137: Lump this one in with the iconography problem

Issue 144: That sounds like very broken HTML semantics to me. A div that is being used as a dropdown, perhaps. 


Although all this seems like a lot, targeted fixes can fix a reasonable majority of these issues relatively quickly (i.e. fixing one icon can set the path for fixing all of them and that knocks out about 10% of the issues already).
Hope that helps, have a good day you all and thanks for the work that has already been done. Keep up the good work :-)

**Comment by tiffy74** [comment]
Just wow

All I can say is what an amazing and insightful post.  This is some of the most useful information and you've been so thorough and given loads of detail.

All I can say @zersiax is a heartfelt #h5yr and I really hope you can engage more with us as this project develops.  There's so much to do and without posts such as yours, would mean that we may either misunderstand certain accessibility issues or we come up with fixes that actually give bad user experiences for those that they are intended.

Hoping at some point that we'll have a catchup hangout soon and perhaps if you could attend and join in the conversation and provide any insight that could help on certain issues - but if you don't want to that's fine too - you've given loads of great info and I hope you'll help develop this project and give more!  :)

**Comment by nul800sebastiaan** [comment]
Sorry for the close/reopen/close/reopen.. unfortunately when I merge a PR this auto-closes. Wonderful functionality... MOST of the time! 😉😂

**Comment by MMasey** [comment]
PR for 6, 7 and maybe a few other at #5729

**Comment by nul800sebastiaan** [comment]
PR #5729 merged, super nice to be able to use the keyboard in the tree! 🎉

**Comment by BatJan** [comment]
@MMasey Can you update the table with a reference to PR #4526 for items 10,11 and 12 since this, if accepted, will be reusable for these scenarios (If I remember correctly, it's been a while since I made it 😁 ) - But just so people are aware that there is work being done in this area? :)

**Comment by BatJan** [comment]
I have just had a short play with Umbraco 8.1 and it appears that the keyboard navigation issue mentioned for **issue 5** has been fixed - However I'm currently working on a related PR for this issue that should ensure screen reader accessibility and improve semantics too in - It's currently a draft that can be found here https://github.com/umbraco/Umbraco-CMS/pull/5884

Danny Lancaster and I have been chatting a bit about this during the weekend too - Since perhaps this popout should perhaps have the role="menu" and role="menuitem" set since what's being revealed is not links pointing to a url but to some publishing functionality... But I'm awaiting Danny's input and thoughts on this. Of course anyone else is more than welcome to provide me with some input too 😃 

Apart from that what I have currently done is to fix some semantics changing `<a>` to `<button>` since the `<a>` tags did not have any href values. I have also added a visuallyhidden text saying "More publishing options" in the span that is used for showing the caret, which hopefully provides some better context about what the list contains and what happens if it's clicked. I've also added aria-hidden around the ellipsis  after each of the text/labels in the options that appear since they don't really make sense in a non-visual context.

**Comment by BatJan** [comment]
I have created a trello board adding all of the issues and setting some different states in an attempt to create a better overview of what needs to be done, what's in progress, what needs to be reviewed and what has been done. Hopefully it will also be easier to discuss specific issues on this board that it is in this big issue. The board is created here https://trello.com/b/AkmxU8xX/umbraco-accessibility-issues

**Comment by BatJan** [comment]
PR for issue 122 here #5905

**Comment by BatJan** [comment]
PR for issue 99 here https://github.com/umbraco/Umbraco-CMS/pull/5928

**Comment by BatJan** [comment]
PR for issue 104 here https://github.com/umbraco/Umbraco-CMS/pull/5930

**Comment by BatJan** [comment]
PR for issue 88 here https://github.com/umbraco/Umbraco-CMS/pull/5944

**Comment by BatJan** [comment]
PR for issue 86 here #5960

**Comment by BatJan** [comment]
PR for issues 62,82 and 87 here https://github.com/umbraco/Umbraco-CMS/pull/5963

**Comment by BatJan** [comment]
Pr for issue 11 here https://github.com/umbraco/Umbraco-CMS/pull/5964

**Comment by BatJan** [comment]
Pr for issue 111 here #5970

**Comment by RachBreeze** [comment]
PRs added for issues with titles on login screen and content edit section https://github.com/umbraco/Umbraco-CMS/pull/6236 https://github.com/umbraco/Umbraco-CMS/pull/5888

**Comment by RachBreeze** [comment]
Also failed to mention other accessibility PRs for login screen #5800 and #5807

**Comment by BatJan** [comment]
PR added for issue 8 and 45 here #6303 

PR added for issue 9 here #6301

**Comment by BatJan** [comment]
PR added for issue **34* here #6305 

PR added for issue **37** here #6304

**Comment by BatJan** [comment]
PR added for issue **157** #6402

**Comment by MMasey** [comment]
PR #6804 added for following issues:
- 32 & 60 ( "+" icon is identified as "link" - screen readers)
- 46 & 54 - Once new row has been added to page grid, unable to navigate to using keyboard


P.S. I know the table needs some TLC. Once the fresh 8.2 audit is complete, i'll probably remove it entirely in favour of this trello board. https://trello.com/b/AkmxU8xX/umbraco-accessibility-issues

If anyone has any alternative suggesstions/ideas/reservations please let me know.

**Comment by MMasey** [comment]
Another PR for fixing an issue with the media picker and keyboard accessibility. I think it may be the same at 46, but i'm not 100% sure. #6805

**Comment by umbrabot** [comment]
Hiya @MMasey,

Just wanted to let you know that we noticed that this issue got a bit stale and might not be relevant any more.

We will close this issue for now but we're happy to open it up again if you think it's still relevant (for example: it's a feature request that's not yet implemented, or it's a bug that's not yet been fixed). 

To open it this issue up again, you can write `@umbrabot still relevant` in  a new comment as the first line. It would be *super* helpful for us if on the next line you could let us know why you think it's still relevant.

For example:

>@umbrabot still relevant
>This bug can still be reproduced in version `x.y.z` 

This will reopen the issue in the next few hours.

Thanks, from your friendly Umbraco GitHub bot :robot: :slightly_smiling_face:
